File: backend/app/services/matching_engine.py
from typing import List, Dict, Any, Optional

def calculate_opportunity_match(*args, **kwargs) -> dict:
    """
    Authoritative, Multi-Factor, Data-Driven Matching Algorithm:
    Single Source of Truth based on real SQLite database records:
    - Technical Competencies & Skills (50% max)
    - Educational Eligibility (15% max)
    - AYUSH Discipline Alignment (10% max)
    - Career Interest & Sector Fit (10% max)
    - Practical Projects / Experience (5% max)
    - Verified Industry Certifications (5% max)
    - Work Mode / Location Fit (5% max)
    
    Zero Artificial Inflation:
    - Missing mandatory skills severely gatekeep/penalize candidate compatibility.
    - Unmatched dimensions score 0 points, eliminating the artificial 30% floor.
    """
    student_data = kwargs.get("student_data") or kwargs.get("candidate_skills")
    opp_data = kwargs.get("opp_data") or kwargs.get("required_skills") or kwargs.get("opportunity_data")

    if args:
        if len(args) >= 1 and student_data is None:
            student_data = args[0]
        if len(args) >= 2 and opp_data is None:
            opp_data = args[1]

    if isinstance(opp_data, list):
        opp_data = {"skills": opp_data}
    elif not isinstance(opp_data, dict):
        opp_data = {}

    if isinstance(student_data, list):
        student_data = {"skills": student_data}
    elif not isinstance(student_data, dict):
        student_data = {}

    strengths: List[str] = []
    gaps: List[str] = []
    missing_mandatory: List[str] = []
    explanations: List[str] = []

    # 1. Format student skills map
    student_skills_map: Dict[str, float] = {}
    if isinstance(student_data, dict):
        if "skills" in student_data and isinstance(student_data["skills"], list):
            for s in student_data["skills"]:
                if isinstance(s, dict):
                    s_id = str(s.get("skill_id") or s.get("id") or "")
                    if s_id:
                        student_skills_map[s_id] = float(s.get("proficiency_score", s.get("proficiency", 0.0)))
        else:
            for k, v in student_data.items():
                if isinstance(v, (int, float)):
                    student_skills_map[str(k)] = float(v)
                elif isinstance(v, dict):
                    student_skills_map[str(k)] = float(v.get("proficiency_score", 0.0))

    # Format opportunity skills list
    opp_skills = []
    if isinstance(opp_data, dict):
        opp_skills = opp_data.get("skills", [])
    elif isinstance(opp_data, list):
        opp_skills = opp_data

    # ----------------------------------------------------
    # 1. TECHNICAL SKILLS MATCH (50% WEIGHT)
    # ----------------------------------------------------
    skill_score = 0.0
    total_mandatory_skills = 0
    missing_mandatory_count = 0

    if not opp_skills:
        # General opportunity with no specialized skills required
        skill_score = 25.0
        explanations.append("Open eligibility: foundational competency scope.")
    else:
        total_weight = sum(float(s.get("weight", 1.0)) for s in opp_skills) or 1.0
        weighted_sum = 0.0

        for os in opp_skills:
            s_id = str(os.get("skill_id") or os.get("id") or "")
            req_prof = float(os.get("required_proficiency", 60.0))
            weight = float(os.get("weight", 1.0))
            is_mandatory = bool(os.get("is_mandatory", True))
            skill_name = str(os.get("skill_name") or os.get("name") or "Required Competency")

            if is_mandatory:
                total_mandatory_skills += 1

            if s_id in student_skills_map:
                prof = student_skills_map[s_id]
                ratio = min(1.0, prof / max(req_prof, 1.0))

                # Penalty if mandatory and severely deficient (< 50% of benchmark)
                if is_mandatory and prof < req_prof * 0.5:
                    ratio *= 0.75
                    missing_mandatory_count += 1
                    missing_mandatory.append(skill_name)
                    gaps.append(f"Critical deficit in mandatory skill: {skill_name} ({int(prof)}% vs required {int(req_prof)}%)")
                elif prof >= req_prof:
                    strengths.append(f"Exceeds {skill_name} requirement ({int(prof)}% vs {int(req_prof)}%)")
                else:
                    gaps.append(f"Moderate deficit in {skill_name} ({int(prof)}% vs required {int(req_prof)}%)")

                weighted_sum += ratio * weight
            else:
                # Skill completely unassessed
                if is_mandatory:
                    missing_mandatory_count += 1
                    missing_mandatory.append(skill_name)
                    gaps.append(f"Missing mandatory skill: {skill_name} (Not Assessed)")
                else:
                    gaps.append(f"Unassessed recommended skill: {skill_name}")

        normalized_ratio = weighted_sum / total_weight
        skill_score = round(normalized_ratio * 50.0, 1)

    # ----------------------------------------------------
    # 2. EDUCATION MATCH (15% WEIGHT)
    # ----------------------------------------------------
    education_score = 0.0
    degree_req = str(opp_data.get("eligibility_degree") or "").upper().strip()
    student_degree = str(student_data.get("degree") or "").upper().strip()

    if not degree_req or "ANY" in degree_req:
        education_score = 15.0
        strengths.append("Educational eligibility satisfied (Open requirement)")
    elif student_degree:
        # Check standard AYUSH degrees
        ayush_degrees = ["BAMS", "BHMS", "BUMS", "BSMS", "BNYS", "MD", "MS", "PHD", "M.PHARM", "MPH", "B.TECH"]
        is_ayush_match = any(d in student_degree and d in degree_req for d in ayush_degrees)
        
        if is_ayush_match or student_degree in degree_req or degree_req in student_degree:
            education_score = 15.0
            strengths.append(f"Degree matches requirement: {student_degree}")
        elif "AYUSH" in degree_req and any(d in student_degree for d in ["BAMS", "BHMS", "BUMS", "BSMS", "BNYS"]):
            education_score = 15.0
            strengths.append(f"AYUSH degree satisfies criteria: {student_degree}")
        else:
            education_score = 0.0
            gaps.append(f"Degree mismatch: holds {student_degree} but position requires {degree_req}")
    else:
        # No degree on record
        education_score = 0.0
        gaps.append("Educational degree not specified in student profile")

    # ----------------------------------------------------
    # 3. AYUSH DISCIPLINE MATCH (10% WEIGHT)
    # ----------------------------------------------------
    discipline_score = 0.0
    opp_disc_id = opp_data.get("discipline_id")
    student_disc_id = student_data.get("ayush_discipline_id")

    if not opp_disc_id:
        discipline_score = 10.0
        strengths.append("Multi-disciplinary AYUSH opportunity")
    elif student_disc_id and opp_disc_id == student_disc_id:
        discipline_score = 10.0
        strengths.append("Direct AYUSH discipline alignment")
    else:
        # Unrelated discipline gives 0 points
        discipline_score = 0.0
        gaps.append("Discipline does not align with targeted domain")

    # ----------------------------------------------------
    # 4. CAREER SECTOR FIT (10% WEIGHT)
    # ----------------------------------------------------
    career_score = 0.0
    opp_sector_id = opp_data.get("sector_id")
    student_sector_id = (
        student_data.get("target_career_sector_id") 
        or student_data.get("sector_id")
        or (student_data.get("target_career_role") or {}).get("sector_id")
    )
    student_role_id = student_data.get("target_career_role_id")

    if opp_sector_id and student_sector_id and opp_sector_id == student_sector_id:
        career_score = 10.0
        strengths.append("Direct career sector alignment")
    elif student_role_id:
        career_score = 5.0
    else:
        career_score = 0.0

    # ----------------------------------------------------
    # 5. PRACTICAL PROJECTS (5% WEIGHT)
    # ----------------------------------------------------
    p_count = int(student_data.get("projects_count", 0))
    if p_count >= 2:
        experience_score = 5.0
        strengths.append(f"Proven project portfolio ({p_count} projects)")
    elif p_count == 1:
        experience_score = 2.5
        strengths.append("Practical project experience demonstrated")
    else:
        experience_score = 0.0
        gaps.append("No practical projects on record")

    # ----------------------------------------------------
    # 6. VERIFIED CERTIFICATIONS (5% WEIGHT)
    # ----------------------------------------------------
    c_count = int(student_data.get("certifications_count", 0))
    if c_count >= 2:
        cert_score = 5.0
        strengths.append(f"Verified certifications ({c_count} credentials)")
    elif c_count == 1:
        cert_score = 2.5
        strengths.append("Verified certification on record")
    else:
        cert_score = 0.0

    # ----------------------------------------------------
    # 7. WORK MODE & LOCATION (5% WEIGHT)
    # ----------------------------------------------------
    loc_score = 0.0
    work_mode = str(opp_data.get("work_mode") or "").upper()
    opp_loc = str(opp_data.get("location") or "").lower()
    student_loc = str(student_data.get("location") or "").lower()

    if work_mode == "REMOTE":
        loc_score = 5.0
        strengths.append("100% Remote flexibility")
    elif student_loc and opp_loc and (student_loc in opp_loc or opp_loc in student_loc):
        loc_score = 5.0
        strengths.append(f"Location match ({opp_data.get('location')})")
    elif work_mode == "HYBRID":
        loc_score = 3.0
    else:
        # On-site with location mismatch
        loc_score = 0.0

    # ----------------------------------------------------
    # OVERALL CALCULATION & MANDATORY GATEKEEPING
    # ----------------------------------------------------
    raw_total = (
        skill_score + 
        education_score + 
        discipline_score + 
        career_score + 
        experience_score + 
        cert_score + 
        loc_score
    )

    # Mandatory gatekeeper penalty:
    # If student is missing mandatory skills, they cannot pass as a high match.
    if missing_mandatory_count > 0:
        penalty_ratio = min(1.0, missing_mandatory_count / max(total_mandatory_skills, 1))
        # Reduce score up to 60% based on proportion of missing mandatory competencies
        penalty_factor = max(0.1, 1.0 - (penalty_ratio * 0.65))
        raw_total *= penalty_factor
        explanations.append(f"Score reduced due to {missing_mandatory_count} missing mandatory competencies.")

    overall = min(100.0, max(0.0, round(raw_total, 1)))

    return {
        "overall_score": overall,
        "match_score": overall,
        "skill_score": skill_score,
        "education_score": education_score,
        "discipline_score": discipline_score,
        "career_score": career_score,
        "experience_score": experience_score,
        "certification_score": cert_score,
        "location_score": loc_score,
        "strengths": strengths[:4],
        "gaps": gaps[:4],
        "missing_skills": gaps[:4],
        "missing_mandatory": missing_mandatory,
        "explanations": explanations,
        "breakdown": {
            "skillScore": skill_score,
            "educationScore": education_score,
            "disciplineScore": discipline_score,
            "careerScore": career_score,
            "experienceScore": experience_score,
            "certificationScore": cert_score,
            "locationScore": loc_score
        }
    }

File: frontend/src/lib/services/matchingEngine.ts
// Deterministic & Explainable Matching Engine
// Weights: Technical Skill (50%), Education (15%), Discipline (10%), Career Sector (10%), Experience (5%), Certification (5%), Location (5%)
// Strictly data-driven with Zero Artificial Baseline Inflation and Mandatory Skill Gatekeeping

export interface MatchBreakdown {
  overallScore: number; // 0 - 100
  skillScore: number; // 0 - 50
  educationScore: number; // 0 - 15
  disciplineScore: number; // 0 - 10
  careerInterestScore: number; // 0 - 10
  experienceScore: number; // 0 - 5
  certificationScore: number; // 0 - 5
  locationScore: number; // 0 - 5
  strengths: string[];
  gaps: string[];
  explanations: string[];
  missingMandatory?: string[];
}

export interface StudentMatchProfile {
  id: string;
  degree: string;
  ayushDisciplineId: string;
  ayushDisciplineName?: string;
  targetCareerRoleId?: string | null;
  targetCareerSectorId?: string | null;
  location?: string | null;
  preferredWorkMode?: string;
  skills: Array<{
    skillId: string;
    skillName: string;
    proficiencyScore: number;
    verificationLevel?: string;
  }>;
  projectsCount: number;
  certificationsCount: number;
}

export interface OpportunityMatchTarget {
  id: string;
  title: string;
  opportunityType: string;
  sectorId: string;
  disciplineId?: string | null;
  disciplineName?: string;
  eligibilityDegree: string;
  location: string;
  workMode: string; // REMOTE, HYBRID, ONSITE
  skills: Array<{
    skillId: string;
    skillName: string;
    requiredProficiency: number;
    isMandatory: boolean;
    weight?: number;
  }>;
}

export function calculateOpportunityMatch(
  student: StudentMatchProfile,
  opportunity: OpportunityMatchTarget
): MatchBreakdown {
  const strengths: string[] = [];
  const gaps: string[] = [];
  const explanations: string[] = [];
  const missingMandatory: string[] = [];

  // ----------------------------------------------------
  // 1. SKILL MATCH (50% WEIGHT)
  // ----------------------------------------------------
  let skillPoints = 0;
  let totalMandatory = 0;
  let missingMandatoryCount = 0;

  const studentSkillMap = new Map<string, { score: number; level?: string }>();
  for (const s of student.skills || []) {
    studentSkillMap.set(s.skillId, { score: s.proficiencyScore, level: s.verificationLevel });
  }

  if (opportunity.skills && opportunity.skills.length > 0) {
    let weightedRatioSum = 0;
    let totalWeight = 0;

    for (const req of opportunity.skills) {
      if (req.isMandatory) totalMandatory++;

      const studentSkill = studentSkillMap.get(req.skillId);
      const studentScore = studentSkill ? studentSkill.score : 0;
      const requiredScore = req.requiredProficiency || 50;
      const weight = req.weight || 1.0;

      totalWeight += weight;

      if (!studentSkill || studentScore <= 0) {
        if (req.isMandatory) {
          missingMandatoryCount++;
          missingMandatory.push(req.skillName);
          gaps.push(`Missing mandatory skill: ${req.skillName} (Not Assessed)`);
        } else {
          gaps.push(`Unassessed recommended skill: ${req.skillName}`);
        }
      } else {
        let ratio = Math.min(1.0, studentScore / requiredScore);

        if (req.isMandatory && studentScore < requiredScore * 0.5) {
          ratio *= 0.75;
          missingMandatoryCount++;
          missingMandatory.push(req.skillName);
          gaps.push(`Critical deficit in mandatory skill: ${req.skillName} (${Math.round(studentScore)}% vs required ${Math.round(requiredScore)}%)`);
        } else if (studentScore >= requiredScore) {
          strengths.push(`Strong proficiency in ${req.skillName} (${Math.round(studentScore)}% vs required ${Math.round(requiredScore)}%)`);
        } else {
          gaps.push(`Moderate deficit in ${req.skillName} (${Math.round(studentScore)}% vs required ${Math.round(requiredScore)}%)`);
        }

        weightedRatioSum += ratio * weight;
      }
    }

    const avgSkillRatio = totalWeight > 0 ? weightedRatioSum / totalWeight : 0;
    skillPoints = Math.round(avgSkillRatio * 50 * 10) / 10;
  } else {
    // Open skills opportunity
    skillPoints = 25;
    strengths.push("Foundational skills align with opportunity scope");
  }

  // ----------------------------------------------------
  // 2. EDUCATION MATCH (15% WEIGHT)
  // ----------------------------------------------------
  let educationPoints = 0;
  const reqDegrees = (opportunity.eligibilityDegree || "").toLowerCase().trim();
  const studentDegree = (student.degree || "").toLowerCase().trim();

  if (!reqDegrees || reqDegrees.includes("any")) {
    educationPoints = 15;
    strengths.push("Educational eligibility open to all degrees");
  } else if (studentDegree) {
    const degreesList = ["bams", "bhms", "bums", "bsms", "bnys", "mpharm", "msc", "mph", "btech", "md", "ms", "phd"];
    const matchesExact = degreesList.some(
      (d) => studentDegree.includes(d) && reqDegrees.includes(d)
    );

    if (matchesExact || reqDegrees.includes(studentDegree) || studentDegree.includes(reqDegrees)) {
      educationPoints = 15;
      strengths.push(`Degree matches requirement: ${student.degree}`);
    } else if (reqDegrees.includes("ayush") && degreesList.some((d) => studentDegree.includes(d))) {
      educationPoints = 15;
      strengths.push(`AYUSH medical degree satisfies eligibility: ${student.degree}`);
    } else {
      educationPoints = 0;
      gaps.push(`Degree mismatch: holds ${student.degree} but requires ${opportunity.eligibilityDegree}`);
    }
  } else {
    educationPoints = 0;
    gaps.push("Educational degree not specified in student profile");
  }

  // ----------------------------------------------------
  // 3. AYUSH DISCIPLINE MATCH (10% WEIGHT)
  // ----------------------------------------------------
  let disciplinePoints = 0;
  if (!opportunity.disciplineId) {
    disciplinePoints = 10; // Open to all AYUSH disciplines
    strengths.push("Open to all AYUSH disciplines");
  } else if (student.ayushDisciplineId && opportunity.disciplineId === student.ayushDisciplineId) {
    disciplinePoints = 10;
    strengths.push(`AYUSH discipline matches: ${student.ayushDisciplineName || "Discipline"}`);
  } else {
    disciplinePoints = 0;
    gaps.push("Discipline does not align with targeted domain");
  }

  // ----------------------------------------------------
  // 4. CAREER SECTOR FIT (10% WEIGHT)
  // ----------------------------------------------------
  let careerPoints = 0;
  if (student.targetCareerSectorId && opportunity.sectorId && student.targetCareerSectorId === opportunity.sectorId) {
    careerPoints = 10;
    strengths.push("Direct career sector alignment");
  } else if (student.targetCareerRoleId) {
    careerPoints = 5.0;
  } else {
    careerPoints = 0;
  }

  // ----------------------------------------------------
  // 5. EXPERIENCE MATCH (5% WEIGHT)
  // ----------------------------------------------------
  let experiencePoints = 0;
  if (student.projectsCount >= 2) {
    experiencePoints = 5;
    strengths.push(`Proven project portfolio (${student.projectsCount} projects)`);
  } else if (student.projectsCount === 1) {
    experiencePoints = 2.5;
    strengths.push("Practical project experience demonstrated");
  } else {
    experiencePoints = 0;
    gaps.push("No practical projects on record");
  }

  // ----------------------------------------------------
  // 6. CERTIFICATION MATCH (5% WEIGHT)
  // ----------------------------------------------------
  let certificationPoints = 0;
  if (student.certificationsCount >= 2) {
    certificationPoints = 5;
    strengths.push("Multiple verified certifications in profile");
  } else if (student.certificationsCount === 1) {
    certificationPoints = 2.5;
    strengths.push("Verified certification on record");
  } else {
    certificationPoints = 0;
  }

  // ----------------------------------------------------
  // 7. LOCATION & WORK MODE MATCH (5% WEIGHT)
  // ----------------------------------------------------
  let locationPoints = 0;
  const oppMode = (opportunity.workMode || "").toUpperCase();
  const oppLoc = (opportunity.location || "").toLowerCase();
  const studLoc = (student.location || "").toLowerCase();

  if (oppMode === "REMOTE") {
    locationPoints = 5;
    strengths.push("100% Remote flexibility");
  } else if (studLoc && oppLoc && (oppLoc.includes(studLoc) || studLoc.includes(oppLoc))) {
    locationPoints = 5;
    strengths.push(`Location matches: ${opportunity.location}`);
  } else if (oppMode === "HYBRID") {
    locationPoints = 3;
  } else {
    locationPoints = 0;
  }

  // ----------------------------------------------------
  // TOTAL SCORE & MANDATORY GATEKEEPER
  // ----------------------------------------------------
  let rawOverall =
    skillPoints +
    educationPoints +
    disciplinePoints +
    careerPoints +
    experiencePoints +
    certificationPoints +
    locationPoints;

  // Gatekeeper penalty: Missing mandatory competencies severely penalizes compatibility
  if (missingMandatoryCount > 0) {
    const penaltyRatio = Math.min(1.0, missingMandatoryCount / Math.max(totalMandatory, 1));
    const penaltyFactor = Math.max(0.1, 1.0 - (penaltyRatio * 0.65));
    rawOverall *= penaltyFactor;
    explanations.push(`Score scaled down due to ${missingMandatoryCount} missing mandatory competencies.`);
  }

  const overallScore = Math.min(100, Math.max(0, Math.round(rawOverall)));

  return {
    overallScore,
    skillScore: Math.round(skillPoints * 10) / 10,
    educationScore: Math.round(educationPoints * 10) / 10,
    disciplineScore: Math.round(disciplinePoints * 10) / 10,
    careerInterestScore: Math.round(careerPoints * 10) / 10,
    experienceScore: Math.round(experiencePoints * 10) / 10,
    certificationScore: Math.round(certificationPoints * 10) / 10,
    locationScore: Math.round(locationPoints * 10) / 10,
    strengths,
    gaps,
    explanations,
    missingMandatory,
  };
}
